core: support arrival on stop signal for conflict detection

bougue-pe opened the pull request:
TODO:
- [x] rebase on TSv2's work (conflicts)
- [x] consider first comments (in a separate commit) ?
- [x] branch on TSv1 (on_stop_signal = true if stop_for is not None) + manual test ? > Not targetted
- [x] port work for TSv2 (copy code, use actual flag from editoast) + manual test ?
- [ ] ~update tests to cover cases imagined in https://github.com/OpenRailAssociation/osrd/issues/7044 and https://github.com/OpenRailAssociation/osrd/issues/7045, check also if these tests are triggering TSv2's code.~ -> will be done in a subsequent PR
- [x] decide whether adding e2e test is a good idea (probably only integration tests).

woshilapin committed:
core: test arrival on stop signal support

So far, there is a test that is supposed to provide a situation where:
- train A goes from SA5 to SD2 through SC5, with a stop in SC5
- train B goes from SA5 to SD2 through SC4, with no stop but leaving after train A
- the full path is clamped at SC1 for train A to avoid a conflict on PC0
- the full path is clamped at SC0 for train B to avoid a conflict on PC0

See https://osrd.fr/en/docs/explanation/models/data-models-full-example/svg_diagrams/small_infra_op_points.drawio.en.svg
for a reference of the infrastructure.

woshilapin commented on core/src/main/java/fr/sncf/osrd/conflicts/IncrementalPath.kt:
```diff
@@ -43,9 +48,10 @@ interface IncrementalPath {
-    val endZonePathIndex: Int
-    val endBlockIndex: Int
-    val endRouteIndex: Int
+    val zonePathCount: Int
+    val blockCount: Int
+    val routeCount: Int
+    val stopCount: Int
```
Seems like a renaming which would have been nice in a different commit?

woshilapin replied:
Done in 99cb207cbf4fa7ef1ab4c7ade2ff70283ba3ff5f.

woshilapin commented on core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.kt:
```diff
@@ -345,14 +346,31 @@ private fun routingRequirements(
+            val stop = stops[stopIdx]
+            val stopTravelledOffset = offsetConverter.toTravelledPath(stop.pathOffset)
+            if (stop.onStopSignal && entrySignalOffset <= stopTravelledOffset) {
+                criticalPos = stopTravelledOffset + Distance(1)
```
Would be nice to comment here, because it's not really clear. From what I understand, `criticalPos` helps projecting in time, so putting just after the stop makes it so it already waited for the stop to be finished and now can reserve the route/blocks in front of him (if it does that before the stop, then the route would be reserved for the time of the stop which is not necessary). Is that right?

Or maybe we could `return` from here (instead of `break`) with `envelope.clampInterpolate(criticalPos) + stop.duration` (we might need to add the duration in `PathStop`). I believe it would be semantically correct (instead of this hack of adding a millimeter).

bougue-pe replied:
This is right, the additional 1 millimeter is used to interpolate (hack/approximation) departure time from stop.
The absolute correct use would be that `clampInterpolate` from position to be able to return a (multiple?) period and use the final datetime in that case.

A comment was added in 2 places: https://github.com/OpenRailAssociation/osrd/pull/7322/files#diff-a682d3252344adbc76cac3d4e7014e8b786ccef102a2eb32367744294384fea3R42-R43 and https://github.com/OpenRailAssociation/osrd/pull/7322/files#diff-8078a2f57f53541732c3008fa2e15551402baba5959c846d63fa84e9d9637e4bR371-R372

woshilapin commented on core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.kt:
```diff
@@ -328,21 +346,38 @@ private fun routingRequirements(
+        }
 
         // find when the train meets the critical location
         return envelope.clampInterpolate(criticalPos)
```
If I understand correctly, the API of this `clampInterpolate` is a bit weird, it always returns one value, but in case of a stop, it should return a period (or a start value and a duration): a simple position can be projected on multiple times. Isn't it?

bougue-pe replied:
see https://github.com/OpenRailAssociation/osrd/pull/7322#discussion_r1620295815

woshilapin commented on core/src/main/java/fr/sncf/osrd/conflicts/IncrementalRequirementEnvelopeAdapter.kt:
```diff
@@ -16,17 +17,35 @@ class IncrementalRequirementEnvelopeAdapter(
+        if (stopOffset.distance.meters > endPos) {
+            return Double.POSITIVE_INFINITY
+        }
+        var pastStop = (stopOffset.distance + Distance(1)).meters
```
Can we do better than this hack `Distance(1)`? It's not so obvious why we do that and at least a comment would help.

bougue-pe replied:
Done, see https://github.com/OpenRailAssociation/osrd/pull/7322#discussion_r1620295815

woshilapin commented on core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt:
```diff
@@ -293,6 +293,7 @@ private class InfraExplorerImpl(
                         mutableStaticIdxArrayListOf(block),
                         containsStart = startPath,
                         containsEnd = endPath,
+                        stops = listOf(),
```
Why do we send an empty list here? Is there a reason? If yes, it probably needs a comment why.

woshilapin replied:
After discussion, it's a left over (the PR is still a WIP). So yes, there is something missing here. To be defined.

woshilapin replied:
Done in https://github.com/OpenRailAssociation/osrd/pull/7322/commits/88e815c5bb774f1b1a150ea1cf5d14c95d6ad2e0.

woshilapin commented on core/src/main/java/fr/sncf/osrd/conflicts/SpacingResourceGenerator.kt:
```diff
@@ -57,15 +63,18 @@ class SpacingRequirementAutomaton(
     // requirements that need to be returned on the next successful pathUpdate
     private val pendingRequirements = ArrayDeque<PendingSpacingRequirement>()
 
+    private var nextProcessedStop = 0
```
I believe it should be named `nextStopToProcess` or `nextProcessedStopCandidate` because it's not about already processed stops if I understand correctly.

woshilapin replied:
Done in 2c8f101788c0e3fecb68620974c7c0a2f82e34b3.

woshilapin commented on core/src/main/java/fr/sncf/osrd/conflicts/SpacingResourceGenerator.kt:
```diff
@@ -330,34 +359,62 @@ class SpacingRequirementAutomaton(
+        val spacingRequirements =
+            pendingRequirements
+                .mapIndexed { index, pendingSpacingRequirement ->
+                    val serializedReq = postProcessRequirement(pendingSpacingRequirement)
```
`serializedReq` might be renamed too into `spacingRequirement`.

woshilapin replied:
Done in ce71668a1844887407e778be15b9196a359eea2d.

woshilapin reviewed (commented):
Just a first pass of comments for the moment, trying to make sense of all of this.

woshilapin committed:
core: renaming of counts (not using index)

woshilapin committed:
core, editoast: support arrival-on-stop-signal for conflict detection

Done for spacing and routing requirements
Main target is TrainScheduleV2 (STDCM and TSv1 only minimal)
Update tests accordingly

Also:
* add PathOffsetBuilder between Path and TravelledPath
* fix offset types of ZoneUpdate and SignalSighting
* get rid of test cases with empty stops in InfraExplorer

Co-authored-by: Younes Khoudli <younes.khoudli@epita.fr>
Co-authored-by: Victor "multun" Collod <victor.collod@epita.fr>
Co-authored-by: Pierre-Etienne Bougué <bougue.pe@proton.me>

bougue-pe committed:
tests: add integration tests for arrival on stop signal

flomonster reviewed (approved):
Reviewed Editoast only

eckter commented on core/src/main/java/fr/sncf/osrd/conflicts/IncrementalRequirementEnvelopeAdapter.kt:
```diff
@@ -16,17 +17,37 @@ class IncrementalRequirementEnvelopeAdapter(
+        if (stopOffset.distance.meters > endPos) {
+            return Double.POSITIVE_INFINITY
+        }
+        // (stop offset + 1 millimeter) is the first position after stop, used to approximate/hack
+        // stop-departure time
```
It's a bit of a shame to need this trick, it's the `envelopeWithStops.interpolateTotalTime()` call just below that iterates over the stops to include their durations. We could imagine an envelope method to get time at a stop location that lets us input if we want the time before or after the stop.

But that's not much of a problem considering the envelopes shouldn't stay around for very long. 

Edit, wait actually, 1mm is weird here. It's either not enough or not required altogether. When adding the stop duration in `EnvelopeStopWrapper.interpolateTotalTime`, we use `arePositionsEqual` which has a 1cm tolerance. It seems to skip the stop if we're within 1cm of the stop location.

It *seems* to work according to the tests, but there's a lot of red flags and *potential* bugs around there.

Edit²: I misread the content of `EnvelopeStopWrapper.interpolateTotalTime`, it counts the stop time when getting a time near a stop (+- 1cm). So it does work and the added 1mm is unnecessary. But it relies on a behavior that isn't entirely obvious.

bougue-pe replied:
After discussion:
The current code of `interpolateTotalTime` currently returns the end-time of a stop.
Which means that this 1mm hack is useless (but doesn't change behavior as there is a 1cm epsilon).

The interface of `EnveloppeInterpolate` could be improved to let the choice to get stop-start or stop-end.
Adding this is reasonnably doable but raises a few questions: so not doing it in this PR, but removing the useless 1mm hack (and adding comments).

bougue-pe replied:
Note that the doc on `InterpolateEnveloppe` is currently misleading or wrong, but I'm not digging this for now (same questions on which method(s) is/are responsible for default behavior).

eckter commented on core/src/main/java/fr/sncf/osrd/conflicts/SpacingResourceGenerator.kt:
```diff
@@ -57,15 +63,18 @@ class SpacingRequirementAutomaton(
     // requirements that need to be returned on the next successful pathUpdate
     private val pendingRequirements = ArrayDeque<PendingSpacingRequirement>()
 
+    private var nextStopToProcess = 0
+    private var processedStops = ArrayDeque<ProcessedStop>()
+
```
Should most likely be included in `clone()`

bougue-pe replied:
Done in https://github.com/OpenRailAssociation/osrd/pull/7322/commits/1c9bce5bfc2c52375414a9b6f25297b7dba021e5

eckter commented on core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt:
```diff
@@ -145,8 +148,12 @@ private class InfraExplorerImpl(
     private var currentIndex: Int = 0,
-    private val endBlocks:
-        Collection<BlockId>, // Blocks on which "end of path" should be set to true
+    // TODO: Should evolve into a List of Collection of struct that contains location, duration and
+    // onStopSignal
```
```suggestion
    // TODO: Should evolve into a List of struct that contains duration, onStopSignal,
    // and a collection of locations
```

The collection only applies to the locations

bougue-pe replied:
Done in https://github.com/OpenRailAssociation/osrd/pull/7322/commits/1c9bce5bfc2c52375414a9b6f25297b7dba021e5

eckter commented on core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt:
```diff
@@ -145,8 +148,12 @@ private class InfraExplorerImpl(
+    private val stops:
+        List<
+            Collection<PathfindingEdgeLocationId<Block>>
+        >, // Blocks on which "end of path" should be set to true
```
```suggestion
        >,
```

Outdated

bougue-pe replied:
Done in https://github.com/OpenRailAssociation/osrd/pull/7322/commits/1c9bce5bfc2c52375414a9b6f25297b7dba021e5

eckter commented on core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt:
```diff
@@ -286,19 +293,36 @@ private class InfraExplorerImpl(
                 if (isRouteBlocked) return false
-                val endPath = endBlocks.contains(block)
+                val endLocationsOnBlock =
+                    if (stops.isEmpty()) {
+                        listOf()
```
It feels like this should be impossible. Has this case happened during testing?

I don't think we would find any valid path in this case, we could just `return false` to avoid the pathfinding process entirely

bougue-pe replied:
It's indeed happening during tests (17 cases).
Returning `false` also crashes, so it's debatable, but I'd keep it like that as it doesn't degrade things.
(adding a comment as suggested by @Khoyo)

eckter replied:
Oh I forgot that those aren't just any waypoints but specifically points where the train stops.

This means that the last waypoint isn't always a stop. This was supposed to be enforced in the stdcm endpoint though an assert, but maybe this case happens in unit tests? 

The problem here is that we use the last *stop* to define when the path is marked as complete, when we should be using the last waypoint. This will cause bugs, the kind that may go unnoticed but are still problematic.

The easiest way to handle this would be to *ensure* that the last waypoint is a stop, including in tests (though either asserts or by setting it by hand). The other solution would be to keep the previous parameter (`endBlocks`) on top of the stops.

bougue-pe replied:
As discussed, we dug a cleaner solution in https://github.com/OpenRailAssociation/osrd/pull/7322/commits/d56a4c1a43aaf050ee07f9c47a1e36b34cd1cca3

eckter reviewed (commented):
LGTM for tests (which is where my review was actually required), but there's some weird stuff going on in core. Mostly minor comments but there's one that requires careful double-checking
